### PR TITLE
[REG-1525] A whole new world of state

### PR DESCRIPTION
**Pull request by RG-nAmKcAz**
What is not implemented yet
- JS Remote bots have not been updated for this yet [Linear Follow Up](https://linear.app/regression-games/issue/REG-1574/update-js-library-and-agentbuilder-js-generation-based-on-new-state)

What this PR removes ...
- The concept of attaching State and/or Action Behaviours to GameObjects. 
- RGEntity ... (sort of replaced with RGStatefulGameObject, but simpler ... also [RGStateType] for specifying isPlayer)
- RGState
- RGAction

What this PR changes ...
- class/package names for State/Action related class hierarchies to avoid confusion with old concepts
- How states/actions are found at runtime
  - BehavioursWithStateOrActions.cs now handles loading mappings of MonoBehaviours that support [RGState] or [RGAction]
  - These mappings are used during runtime by State and Action Handler classes to populate state and handle actions
- How state is represented for each GameObject
  - the 'core' state (position/name/etc) is always at the entry for each gameObject in the state.  Behaviour based actions/states are additional fields under that core object each with their own stateentity value
  - new accessors have been added in both RG.cs and in GetField to make getting this nested values more natural
- How actions are handled
  - At runtime, ActionHandler and StateHandler behaviours are added to any needed gameObjects.
  - When an action is invoked, the RGActionRequest is delegated to the action handler on the appropriate game Objects(s) for that clientId + the overlay for button clicks.
- How/Where our code is generated
  - Generated class formats have changed as everything is about StateEntity or ActionRequest.  The old RGState/RGAction behaviours don't exist anymore.
  - These are all generated into files in the same directory and namespace as their source behaviour with a naming pattern and static field values identifying which behaviour they are paired which which is used to quickly build our mapping at startup.
  - All serialization helpers / etc are generated into the same file needing them to avoid any 1 mega file and conflicts
  - This removes the concept of the generatedscripts folder and removes any conflicts between projects/libraries and their generation.
  - Fixes performance scaling issues we had in generation (generation/extract times are now sub second)

What this PR adds ...
- New class level attribute [RGStateType] which allows specifying a custom typeName for a class, marking that behaviour as `isPlayer`
  - This attribute by default will cause all public fields/properties of that class to be included in the state.
  - If an [RGState] attribute is used in that same class, only those fields/properties/methods with [RGState] will be included
  - What is included can be customized to any combination of NONE,field,property,method
- Additional fields to each entity in the state... all core fields are now
  - id, scene, name, tag, pathInScene, position, rotation, isPlayer, clientId 

Other stuff that got drug into this PR ...
- The platformer 2d state handling stuff was all updated in this PR to get our SMT work up to date.

Updated Documentation
- https://github.com/Regression-Games/RegressionDocs/pull/54

Updated Client (AgentBuilder)
- https://github.com/Regression-Games/RegressionClient/pull/379

Updated Sample Projects
- SDK 3rdPerson Sample in this same project
- https://github.com/Regression-Games/RGBossRoom/pull/46
- https://github.com/Regression-Games/PlatformerSample/pull/6
- https://github.com/Regression-Games/RGCityBuilder/pull/3
- https://github.com/Regression-Games/Match3Sample/pull/5
- SMT : branch `zack/integration_for_follower`

---

Find the pull request instructions [here](https://www.notion.so/regressiongg/Pull-Request-Process-0d57a7eb582a446983edfacafa406f1e?pvs=4)

Every reviewer and the owner of the PR should consider these points in their request (feel free to copy this checklist so you can fill it out yourself in the overall PR comment)

- [ ] The code is extensible and backward compatible
- [ ] New public interfaces are extensible and open to backward compatibility in the future
- [ ] If preparing to remove a field in the future (i.e. this PR removes an argument), the argument stays but is no longer functional, and attaches a deprecation warning. A linear task is also created to track this deletion task.
- [ ] Non-critical or potentially modifiable arguments are optional
- [ ] Breaking changes and the approach to handling them have been verified with the team (in the Linear task, design doc, or PR itself)
- [ ] The code is easy to read
- [ ] Unit tests are added for expected and edge cases
- [ ] Integration tests are added for expected and edge cases
- [ ] Functions and classes are documented
- [ ] Migrations for both up and down operations are completed
- [ ] A documentation PR is created and being reviewed for anything in this PR that requires knowledge to use
- [ ] Implications on other dependent code (i.e. sample games and sample bots) is considered, mentioned, and properly handled
- [ ] Style changes and other non-blocking changes are marked as non-blocking from reviewers

### CODE REVIEW

**Review by analogrelay — APPROVED** [review]
I think this is good to go. I'll be honest, it's quite a lot to review and I probably didn't catch all of it. But I reviewed the generated code and the code paths for executing actions and fetching state in detail.

I have some non-blocking idiomatic feedback.

File: README.md [hunk 1]
```diff
@@ -29,7 +29,7 @@ In the Unity Editor, inside of **Window** > **Package Manager** > **+** >
 **Add package from git URL**, enter this URL (and restart your IDE after adding):
 
 ```
-https://github.com/Regression-Games/RGUnityBots.git?path=src/gg.regression.unity.bots#v0.0.16
+https://github.com/Regression-Games/RGUnityBots.git#v0.0.16
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply README.md:32]: yep, that's a merge oops.. good catch

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs [hunk 2]
```diff
@@ -1,15 +1,26 @@
...
     {
-        public static readonly string HeaderComment = $"/*\r\n* This file has been automatically generated. Do not modify.\r\n*/\r\n\r\n";
-        
+        public static readonly string HeaderComment = "/// <auto-generated>" + Environment.NewLine +
+                                                      "/// This file has been automatically generated. Do not modify." + Environment.NewLine +
+                                                      "/// </auto-generated>" + Environment.NewLine +
+                                                      "// ReSharper disable ClassNeverInstantiated.Global" + Environment.NewLine +
+                                                      "// ReSharper disable RedundantAssignment" + Environment.NewLine +
+                                                      "// ReSharper disable InconsistentNaming" + Environment.NewLine +
+                                                      "// ReSharper disable CheckNamespace" + Environment.NewLine  + Environment.NewLine;
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs:22]: I'm on board with the `all` for these files.  We did a decent but not perfect job generating safe/reliable code.  It might not match every convention, and I'm sure it certainly won't in the future as conventions evolve in newer c# releases.

As for the name, I'm also good with moving the word generated to the 'end' of the name instead of the front where I have it now.


Current

`Generated_RGActions_<SourceBehaviourName>.cs`
`Generated_RGStateEntity_<SourceBehaviourName>.cs`

Proposed

`RGActions_<SourceBehaviourName>_Generated.cs`
`RGStateEntity_<SourceBehaviourName>_Generated.cs`

File: src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+         * <summary>Limit the properties of the call type to be included, Default is 'Field | Property'</summary>
+         */
+        public readonly RGStateIncludeFlags MyFlags;
+        
+        // ReSharper disable once InvalidXmlDocComment
+        /**
+         * <summary>This constructor allows the attribute to specify RGStateIncludeFlags, e.g., [RGStateType(RGStateIncludeFlags.Field)]
+         * Optionally you can also specify whether a game object with this behaviour represents a player avatar object.</summary>
+         */
+        public RGStateTypeAttribute(bool isPlayer = false)
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs:28]: Taking a look / trying it out

File: src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        public enum RGStateIncludeFlags
+        {
+            NONE = 0,
+            // Maybe support limiting visibility of properties later; for now just public
+            // Public = 1,
+            // Serializable = 2,
+            // Private = 4,
+            Field = 16,
+            Property = 32,
+            Method = 64
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs]: updated

File: src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+        
+        //Include Fields and Properties only by default
+        public const RGStateIncludeFlags DefaultFlags = RGStateIncludeFlags.Field | RGStateIncludeFlags.Property;
+
+        /**
+         * <summary>Limit the properties of the call type to be included, Default is 'Field | Property'</summary>
+         */
+        public readonly RGStateIncludeFlags MyFlags;
+        
+        // ReSharper disable once InvalidXmlDocComment
```
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs]: removed/fixed

File: src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/Attributes/RGStateTypeAttribute.cs]: updated to this pattern here and in all the samples

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs]: updated to this new pattern and in all the samples

File: README.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (RG-nAmKcAz) [reply README.md]: resolved

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/GenerateRGActionsClass.cs [hunk 6]
```diff
@@ -0,0 +1,1024 @@
...
+            {
+                if (!botAction.ShouldGenerateCSFile)
+                {
+                    continue;
+                }
+                
+                entityTypeName = botAction.EntityTypeName ?? entityTypeName;
+                
+                // Add RGActionRequest class
+                actionClassDeclarations.Add(
```
Comment (vontell) [inline src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/GenerateRGActionsClass.cs:55]: Is there a preferred way to review this code? It's pretty tricky to follow so I've mostly skimmed and assumed the structure was sound.

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs [hunk 7]
```diff
@@ -107,818 +114,872 @@ public static void GenerateRGScripts()
...
+                // Do NOT include the previous state infos.. so we don't have dupes
+                // This gives us a consistent view across both generated and hand written state class entities
+                EditorUtility.DisplayProgressBar("Extracting Regression Games Agent Builder Data",
+                    "Extracting state info from RGStateEntity classes", 0.2f);
+                var statesInfos = CreateStateInfoFromRGStateEntities();
+                
+                // Do the same for generate RGActionRequest scripts
+                var actionInfos = CreateActionInfoFromRGActionRequests();
+
+                if (_hasExtractProblem)
```
Comment (vontell) [inline src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:142]: Not sure if I am blind, but isn't this always false?

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs [hunk 8]
```diff
@@ -107,818 +114,872 @@ public static void GenerateRGScripts()
...
-
-                // create 'RegressionGames.zip' in project folder
-                EditorUtility.DisplayProgressBar("Extracting Regression Games Agent Builder Data", "Creating .zip file", 0.9f);
-                CreateJsonZip();
-            }
-            finally
-            {
-                EditorUtility.ClearProgressBar();
             }
+            while (stack.Count > 0);
```
Comment (vontell) [inline src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:260]: This may be my GitHub formatting things, but I don't see the body for this?

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs [hunk 9]
```diff
@@ -107,818 +114,872 @@ public static void GenerateRGScripts()
...
+                        }
+                        else if (rgStateTypeAttribute != null)
+                        {
+                            if (includeFlags != RGStateTypeAttribute.RGStateIncludeFlags.NONE)
+                            {
+                                // do logic for all the fields based on the type attribute settings
+                                var includeMembers = classDeclaration.Members
+                                    .Where(m =>
+                                        // if it has [RGState explicitly], or is NOT obsolete
+                                        // iow.. obsolete members come through if annotated with [RGState]
```
Comment (vontell) [inline src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:663]: Is there a reason this should be the case? I think it's fine, just curious. I guess if that are explicitly tagging it, they want it

File: src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs [hunk 10]
```diff
@@ -60,11 +61,11 @@ public override void ProcessTick(RG rgObject)
                 _destination = new Vector3(x, 0, z);
                 
                 // Also, shoot the enemy while we are waiting
-                var enemies = rgObject.FindEntities("Enemy");
+                var enemies = rgObject.FindEntitiesWithType<RGStateEntity_EnemyController>();
                 if (enemies.Count > 0)
                 {
-                    rgObject.PerformAction(new RGActionRequest_MoveInDirection(new Vector2(0, 0)));
-                    rgObject.PerformAction(new RGActionRequest_SelectAndAttackEnemy(enemies[0].id, 0));
+                    rgObject.PerformAction(new RGActionRequest_PlayerInputControl_MoveInDirection(new Vector2(0, 0)));
```
Comment (vontell) [inline src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs:67]: How do we feel about the longer names in bot code? Suppose it's fine and more precise maybe.

File: src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs [hunk 11]
```diff
@@ -49,7 +50,7 @@ public override void ProcessTick(RG rgObject)
             if (thisBot == null) return; // We have not fully spawned yet, so wait
 
             // If we are attacking, just wait, and don't move
-            if (thisBot.GetField<bool>("IsAttacking")) return;
+            if (thisBot.GetFields<RGStateEntity_PlayerAttack>().FirstOrDefault()?.IsAttacking == true) return;
```
Comment (vontell) [inline src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs:53]: This seems more complicated than before? I wonder if there are some utilities and other functions we can add

Comment (vontell) [reply src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs:53]: Actually, if I recall, there was some work on allowing typed access to these - is that the case yet?

File: src/gg.regression.unity.bots/Runtime/Scripts/StateActionTypes/RGStateBehaviour.cs [hunk 12]
```diff
@@ -0,0 +1,44 @@
...
+namespace RegressionGames.StateActionTypes
+{
+    public interface IRGStateBehaviour
+    {
+        public void PopulateStateEntity(RGStateEntity_Core gameObjectCoreState, out bool isPlayer);
+    }
+    
+    public abstract class RGStateBehaviour<T> : MonoBehaviour , IRGStateBehaviour where T : IRGStateEntity
+    {
+        // cache this so we don't create every tick
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/StateActionTypes/RGStateBehaviour.cs:12]: Yeah at some point we should do some official performance testing just to keep track

Comment (vontell) [reply src/gg.regression.unity.bots/Runtime/Scripts/StateActionTypes/RGStateBehaviour.cs:12]: That actually might work well as a test - have a CI/CD pipeline that can compare timing between our SDK and not using the SDK. Def too much work for now but would be cool

File: src/gg.regression.unity.bots/Runtime/Scripts/RGBotConfigs/RGStateProviders/RgStateHandlerPlatformer2DLevel.cs [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+
+        internal Vector3 _lastCellSize = Vector3.one;
+        private List<RgPlatformer2DPosition> _lastPositions = new();
+
+        protected override RgStateEntityBasePlatformer2DLevel CreateStateEntityInstance()
+        {
+            return new RgStateEntityBasePlatformer2DLevel();
+        }
+
+        public override void PopulateStateEntity(RgStateEntityBasePlatformer2DLevel stateEntity)
```
Comment (vontell) [inline src/gg.regression.unity.bots/Runtime/Scripts/RGBotConfigs/RGStateProviders/RgStateHandlerPlatformer2DLevel.cs:81]: Might use this as inspiration for my puzzle experiment since the state there is similar to the cells in a platformer

**Review by vontell — CHANGES_REQUESTED** [review]

File: README.md [hunk 14]
```diff
@@ -29,7 +29,7 @@ In the Unity Editor, inside of **Window** > **Package Manager** > **+** >
 **Add package from git URL**, enter this URL (and restart your IDE after adding):
 
 ```
-https://github.com/Regression-Games/RGUnityBots.git?path=src/gg.regression.unity.bots#v0.0.16
```
[the remark this answers is not in the document]
Comment (vontell) [reply README.md]: This was updated since my review, so resolving

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/GenerateRGActionsClass.cs (hunk 6 above)
Comment (vontell) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/GenerateRGActionsClass.cs:55]: Resolved in DMs

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs (hunk 9 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:663]: My logic.  Obsolete doesn't mean not-supported until they actually remove the field.  They may still want that thing in the state to not break bots until QA can update them.

File: src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs (hunk 11 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs:53]: This is one of the new accessors.

Since every behaviour that has states is now under a field in the state instead of flattened... there is an extra layer of access required.  You can get to that layer by `typeName` (the old way), or by the actual type in C# using this new way.  When you look at the agent builder PR you'll see that it still does things 'the old way', but with this extra layer still accounted for.

File: src/gg.regression.unity.bots/Runtime/Scripts/RGBotConfigs/RGStateProviders/RgStateHandlerPlatformer2DLevel.cs (hunk 13 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Runtime/Scripts/RGBotConfigs/RGStateProviders/RgStateHandlerPlatformer2DLevel.cs:81]: remember to tag these (non-blocking) when applicable

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs (hunk 8 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:260]: this is a do while, you're looking at the end of the loop

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs (hunk 7 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/RGCodeGenerator.cs:142]: `_hasExtractProblem` gets set anytime `RecordError` or `RecordWarning` is called.

These are possibly set at multiple points during the code generation.

File: src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs (hunk 10 above)
Comment (RG-nAmKcAz) [reply src/gg.regression.unity.bots/Samples~/ThirdPersonDemoURP/Bots/QuickstartBot/BotEntryPoint.cs:67]: I don't necessarily love the long names, but it kind of is what it is...

The classnames correlate to the behaviours they go with.
1. For sanity sake so you can easily see what was generated from what
2. You don't have conflicts if the same action name is used in different behaviours

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (analogrelay) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs]: Just a heads-up that if you want this to be detected by tools like Rider as generated code, the file name needs to end with `.Generated.cs` **not** `_Generated.cs`

image

I'm still fine either way though, so this is non-blocking.

**Review by analogrelay — APPROVED** [review]

File: src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (analogrelay) [reply src/gg.regression.unity.bots/Editor/Scripts/CodeGenerators/CodeGeneratorUtils.cs]: Fair enough :)

**Comment by RG-nAmKcAz** [comment]
@analogrelay fyi.. filename pattern has been updated to `.Generated.cs` across all projects :D

**Review by vontell — APPROVED** [review]
